### ISSUE TITLE
New gcc compatibility

### ISSUE THREAD

**Issue by ipatix** [pret/pokeemerald#729]
So I've been doing some attempts to get pokeemerald to compile with gcc 7.3.1. Working with new gcc version will obviously make the code not match anymore, but I'd like to get the codebase to a point where it still matches but requires as few tweaks as possible to use a different gcc version.

I've managed to get it to work mostly, but I want to list a few problems I encountered:

- There are a couple of static variables which follow non static extern declarations in headers. These have to be fixed
- Generally, a few warnings have to go (missing declerations for functions)
- lvalues must not be casted, the code doesn't compile otherwise (see the example below)
- newer gcc will put jump tables into .rodata instead of .text. This causes linking to fail because a lot of files have the .rodata discarded. Perhaps it's just okay to put those .rodata into the linker script as well. If we're lucky .rodata for those files is empty with agbcc and won't cause any memory shifts.
- Apparently inline assembler cannot reference static functions (even if they are in the same file) with newer gccs
- newer gcc will recorder variables. This causes issues in a few places where the code depends on that (example below)
- When compiling with gcc 7, the stack in AgbMain will exceed the 0x40 byte limit, which isn't cleared by RegisterRamReset(RESET_ALL). Because of this the RegisterRamReset call will destory the stack. Change it to RegisterRamReset(0xFD) to fix this.

For lvalues thing:
```
void *foo(void *x, int y)
{
    (int*)x += y;
    return x;
}
```
The code above doesn't compile because you're not allowed to cast lvalues (because x is an lvalue here). Instead you have to do something like this:
```
void *foo(void *x, int y)
{
    *(int **)&x += y;
    return x;
}
```
Because here you just cast the pointer and then dereference it. It seems for basic code that this produces the same code with agbcc. Haven't checked it yet with "problematic" functions. This problem can be found in src/bg.c:CopyToBgTilemapBufferRect

For the reordering issue:
In graphics.c you can find these lines:
```
const u16 gTitleScreenBgPalettes[] = INCBIN_U16("graphics/title_screen/pokemon_logo.gbapal");
const u16 gTitleScreenBgPalettes2[] = INCBIN_U16("graphics/title_screen/rayquaza_and_clouds.gbapal");
```
And in title_screen.c you can find this:
```
LoadPalette(gTitleScreenBgPalettes, 0, 0x1E0);
```
The problem here is that the LoadPalette call assumes those two palette arrays are in order. Apparently since gcc 5 there is an attribute called "no_reorder" which you could possibly use here. Another solution would be to somehow put the two palettes into the same array somehow.

I'm pretty sure issues like this exists in some other places. The problem above caused the title screen palette to get garbled:
![pokeemerald_01](https://user-images.githubusercontent.com/8502545/59677372-f23c4180-91c9-11e9-9d1a-ff555bc43dd8.png)

Other issues I found and haven't figured out their cause:
![pokeemerald-0](https://user-images.githubusercontent.com/8502545/59677319-d173ec00-91c9-11e9-973e-47ae8b1ac14e.png)
This might possibly be related to some memory order thing.

Apart from all of that, the game seems to run fine. That's my progress so far. I'll keep you updated.

**Comment by PikalaxALT** [comment]
See #381 for an earlier attempt at this

**Comment by PikalaxALT** [comment]
For the lvalues thing, your suggested fix is incorrect. Should be ```c
x += y * sizeof(int);
return x:
```

**Comment by huderlem** [comment]
preproc supports the following syntax for incbinning multiple files in order:
```
INCBIN_U16("file1.4bpp", "file2.4bpp");
```
Thought, if you need a reference to the second file, `gTitleScreenBgPalettes2`, this syntax wouldn't allow that.

**Comment by ipatix** [comment]
@huderlem I think in the case of gTitleScreenBgPalettes2, it wasn't referenced at all.
@PikalaxALT Is my version really incorrect? I see that yours makes sense, I'l still have to test if both of them match or don't.

**Comment by SBird1337** [comment]
@ipatix The original source code probably did something along the lines of cating the files together before including them, so its just one array really. To mirror that, the obvious solution should be to only use one array that contains both palettes.

This is done at other places too, but iirc most of the files are already catted with make rules.

**Comment by SBird1337** [comment]
As for fixes that do not result in an OK Rom anymore, you could do something along the lines of
```C
#ifdef MODERN_GCC
...
#else 
...
```

**Comment by PikalaxALT** [comment]
@ipatix 
Oh, I misread your solution, sorry. I withdraw my criticism.

**Comment by ipatix** [comment]
Okay, wasn't hard to guess the cause for the HP bars was the ordering.
https://github.com/pret/pokeemerald/blob/2c22a82a2584355f0710dd97d606e506a6656751/src/graphics.c#L358

Placing an __attribute__((no_reorder)) in front of all the lines fixes the problem.

Just to get an impression: How should we fix this? Should be do this attribute thing with an ifdef (if the compiler version supports it) or should be put it into a single array?

**Comment by PikalaxALT** [comment]
We could do something like this in global.h:
```c
#ifdef MODERN_GCC
#define UNORDERED __attribte__((unordered))
#else
#define UNORDERED
#endif
```

**Comment by ipatix** [comment]
I just added one more comment about an issue with RegisterRamReset destorying the stack on startup.

**Comment by ipatix** [comment]
I just tried the -fno-toplevel-reorder options and it seems to have sorted out all the graphics errors.

**Comment by PikalaxALT** [comment]
> newer gcc will put jump tables into .rodata instead of .text. This causes linking to fail because a lot of files have the .rodata discarded. Perhaps it's just okay to put those .rodata into the linker script as well. If we're lucky .rodata for those files is empty with agbcc and won't cause any memory shifts.

There was one function where I remember this happening, and it broke the `.size` directive on that function (compiler error).  Think it was save.c in Ruby.

**Comment by ipatix** [comment]
Just to give you an overview of what I've done on my end to make everything work with gcc 7:
https://github.com/ipatix/pokeemerald/commit/849c5619089a4d95352146705b8c371ba567abab

**Comment by PikalaxALT** [comment]
> Apparently inline assembler cannot reference static functions (even if they are in the same file) with newer gccs

Resolved in gcc8

**Comment by Diegoisawesome** [comment]
For the reordering stuff, those should ideally be written with the format @huderlem mentioned above, as that is the equivalent of concatenating them and would therefore prevent reordering.

**Comment by PikalaxALT** [comment]
Resolved in #732
